assembly/src/assembler/module_provider.rs

```diff
@@ -0,0 +1,59 @@
+use super::{BTreeMap, Library, LibraryError, Module, ProcedureId, Vec};
+
+// MODULE PROVIDER
+// ================================================================================================
+
+/// A module provider that will store the procedure ASTs used during the compilation process.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ModuleProvider {
+    modules: Vec<Module>,
+    /// Map from procedure id to module index + proc index in the module.
+    procedures: BTreeMap<ProcedureId, usize>,
+}
+
+impl ModuleProvider {
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Fetch a module that contains the provided procedure id.
+    pub fn get_module(&self, id: &ProcedureId) -> Option<&Module> {
+        // this will panic only if there is a bug in `Self::add_module`.
+        self.procedures.get(id).map(|i| &self.modules[*i])
+    }
+
+    // MODULE AND LIBRARY MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Provide a given module.
+    ///
+    /// # Errors
+    ///
+    /// Will error if there is a duplicated module path.
+    pub fn add_module(&mut self, module: Module) -> Result<(), LibraryError> {
+        if self.modules.iter().any(|m| module.path == m.path) {
+            return Err(LibraryError::duplicate_module_path(&module.path));
+        }
+        let module_idx = self.modules.len();
+        module.ast.local_procs.iter().for_each(|proc| {
+            let proc_name = proc.name.to_absolute(&module.path);
+            let proc_id = ProcedureId::from(&proc_name);
+            self.procedures.insert(proc_id, module_idx);
+        });
+        self.modules.push(module);
+        Ok(())
+    }
+
+    /// Provide the modules of a library.
```

Comment:
nit: I would change this to: "Adds all modules from the provided library to this module provider."

Reply:
Fixed!